### PR TITLE
Use Bundler.setup Instead of Bundler.require

### PR DESCRIPTION
**Pull request by gregburek**
http://anti-pattern.com/use-bundler-setup-instead-of-bundler-require

### CODE REVIEW

**Comment by bjeanes** [comment]
YEEEESSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS. 

Thank you. `Bundler.require` is only a concession to rails and was never meant to be the primary interface anyway. 1000 \* :+1:

**Comment by gregburek** [comment]
Some more clean up needs to be done on this PR to get the requires right.

File: lib/app.rb [hunk 1]
```diff
@@ -1,3 +1,11 @@
+require 'pliny'
+require 'sequel'
+require 'honeybadger'
+
+require 'sinatra'
```
Comment (brandur) [inline lib/app.rb:5]: I think we just need `sinatra/base` here.

**Comment by brandur** [comment]
Oh man, thanks for linking that blog post. I can see where he's coming from, but I don't think he's properly acknowledging the fact that apps like these are most often loaded as a single unit. Even if we changed to a model where we started bring things in as one-offs (so every file would have its requires coming in at the top), it's pretty hard to verify that you have the right set everywhere unless you're running every test inside an isolated environment because Ruby treats all requires as a single unified set. I did notice the other day that [Turn provides this functionality](https://github.com/turn-project/turn#command-line), so it might be worth a try.

The tl;dr is that unless we drastically overhaul our approach to dependency loading, I have a feeling that the only practical effect of moving to `.setup` will be that we write more `requires` and Pliny-based projects have more merge conflicts when pulling from upstream.

That said, I'm up for giving it a shot if everyone else thinks it's a good idea! /cc @pedro

**Comment by bjeanes** [comment]
It's easier to combine that to split. Even if the app is loaded mostly as a single unit, that's not in of itself justification for having pieces be able to be operated in isolation. Indeed, Ruby does us no favours in helping keep our dependency set accurate. 

Testing via process isolation (something I'm doing in a few projects — though haven't tried `turn` yet) can combat this but it isn't necessary in order to get value from consciously having explicit dependencies. The cost to having an extra require or a missing require is low (since, as you said, you are mostly loading it as a whole unit) and those issues are quickly caught if you have on average attempted to be explicit with dependencies. 

Missing dependencies are quickly caught when running something in isolation. Extra dependencies are a problem in all projects (I once cleared out 25 unused gems from addons.git!) but having them generally close to their consumers reduces the size of the haystack that has to be searched to vet such extra dependencies.

**Comment by bjeanes** [comment]
(FWIW this is almost entirely the only reason I didn't start with pliny on a recent new api project)

**Comment by bjeanes** [comment]
Oh and lastly, another way to find missing requires is to have an eager loader option that you use when booting it as as a single unit:

``` ruby
app_files = Dir['./lib/**/*.rb'] # or whatever
app_files.shuffle! if ENV.fetch('RACK_ENV', 'development') != 'production'
app_files.map { |f| require f }
```

I've only tried this once so I offer it with a grain of salt. It hasn't bit me so far, as long as I am cognizant of my dependency relationships — which as far as I am concerned is entirely the point.

**Comment by brandur** [comment]
Cool, yeah I'm definitely up for giving it a shot. It seems that to start moving in that direction though, we should probably refactor this pull a bit right? (i.e. move requires closer to where they're used)

**Comment by bjeanes** [comment]
I'll help if Greg doesn't have time.

**Comment by gregburek** [comment]
I think the refactor makes sense here. 

At the moment, I am just working on things as they come up as I am basing a new project off of this. Due to my hacking a rough path through problems, I would appreciate a more experienced ruby eye to help push these requires to where they are needed.  

Also, it will become easier to find dependencies once pliny is fully gem-ified and the code entrance paths settle down. 

For my purposes, using lib/app.rb as the primary code entrance point is working ok, but I will soon be writing unit tests that will need closer requires. @bjeanes: You are welcome to start pushing the requires down until I get there myself.

**Comment by pedro** [comment]
Hey folks! Closing for now as we've moved the template back to Pliny, and pliny-template is no longer in use.
Please reopen in Pliny if you want to keep this going!
Thanks,
